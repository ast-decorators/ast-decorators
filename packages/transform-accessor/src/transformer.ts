import {
  ASTDecoratorDetector,
  ASTDecoratorTransformer,
} from '@ast-decorators/typings';
import minimatch from 'minimatch';
import accessor from './accessor';
import getter from './getter';
import setter from './setter';
import {TransformAccessorOptions} from './utils';

export const TRANSFORMER_NAME = '@ast-decorators/transform-accessor';

const detector = (
  decoratorName: string,
  transformerName: string = TRANSFORMER_NAME,
): ASTDecoratorDetector => (name: string, path: string): boolean =>
  name === decoratorName && minimatch(path, transformerName);

const transformer: ASTDecoratorTransformer = (
  _,
  {transformerPath}: TransformAccessorOptions = {},
) => [
  [accessor, detector('accessor', transformerPath)] as const,
  [getter, detector('getter', transformerPath)] as const,
  [setter, detector('setter', transformerPath)] as const,
];

export default transformer;
