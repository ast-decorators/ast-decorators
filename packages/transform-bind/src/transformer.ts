import {
  ASTDecoratorDetector,
  ASTDecoratorTransformer,
} from '@ast-decorators/utils/lib/common';
import minimatch from 'minimatch';
import bind from './bind';
import bindAll from './bindAll';
import {TransformBindOptions} from './utils';

export const TRANSFORMER_NAME = '@ast-decorators/transform-bind';

const detector = (
  decoratorName: string,
  transformerName: string = TRANSFORMER_NAME,
): ASTDecoratorDetector => (name: string, path: string): boolean =>
  name === decoratorName && minimatch(path, transformerName);

const transformer: ASTDecoratorTransformer = (
  _,
  {transformerPath}: TransformBindOptions = {},
) => [
  [bind, detector('bind', transformerPath)] as const,
  [bindAll, detector('bindAll', transformerPath)] as const,
];

export default transformer;
