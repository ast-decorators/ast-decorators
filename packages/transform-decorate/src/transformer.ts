import type {
  ASTDecoratorDetector,
  ASTDecoratorTransformer,
  ClassMemberMethod,
} from '@ast-decorators/utils/lib/common';
import type {NodePath} from '@babel/core';
import minimatch from 'minimatch';
import {decorateTransformer} from './decorate';
import type {AllowedDecorators, TransformDecorateOptions} from './utils';

export const TRANSFORMER_NAME = '@ast-decorators/transform-decorate';

const detector = (
  decoratorName: string,
  transformerName: string = TRANSFORMER_NAME,
): ASTDecoratorDetector => (name: string, path: string): boolean =>
  name === decoratorName && minimatch(path, transformerName);

const transformer: ASTDecoratorTransformer<
  [NodePath<AllowedDecorators>],
  TransformDecorateOptions,
  ClassMemberMethod
> = (_, {transformerPath}: TransformDecorateOptions = {}) => [
  [decorateTransformer, detector('decorate', transformerPath)] as const,
];

export default transformer;
