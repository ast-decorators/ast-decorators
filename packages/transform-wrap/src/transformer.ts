import type {
  ASTDecoratorDetector,
  ASTDecoratorTransformer,
} from '@ast-decorators/utils/lib/common';
import type {NodePath} from '@babel/core';
import minimatch from 'minimatch';
import {wrapTransformer} from './wrap';
import type {AllowedWrappers, TransformWrapOptions} from './utils';

export const TRANSFORMER_NAME = '@ast-decorators/transform-wrap';

const detector = (
  decoratorName: string,
  transformerName: string = TRANSFORMER_NAME,
): ASTDecoratorDetector => (name: string, path: string): boolean =>
  name === decoratorName && minimatch(path, transformerName);

const transformer: ASTDecoratorTransformer<
  [NodePath<AllowedWrappers>],
  TransformWrapOptions
> = (_, {transformerPath}: TransformWrapOptions = {}) => [
  [wrapTransformer, detector('wrap', transformerPath)] as const,
];

export default transformer;
