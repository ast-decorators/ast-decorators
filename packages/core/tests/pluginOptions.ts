import {
  transform as _transform,
  transformFile as _transformFile,
} from '../../../utils/testing';
import createOptions from './fixtures/plugin-options/options';
import transformer from './fixtures/plugin-options/transformer';

const transform = async (
  fixture: string,
  options: string,
): ReturnType<typeof _transform> =>
  _transform(__dirname, 'plugin-options', fixture, options);

const transformFile = async (
  fixture: string,
  options?: string | object,
): ReturnType<typeof _transformFile> =>
  _transformFile(__dirname, 'plugin-options', fixture, options);

describe('@ast-decorators/core', () => {
  describe('plugin-options', () => {
    it('compiles in one way if option has one state', async () => {
      const {code} = await transformFile(
        'options-first',
        createOptions({
          transformers: [[transformer, {privacy: 'hard'}]],
        }),
      );
      expect(code).toMatchSnapshot();
    });

    it('compiles in another way if option has different state', async () => {
      const {code} = await transformFile(
        'options-second',
        createOptions({
          transformers: [[transformer, {privacy: 'none'}]],
        }),
      );
      expect(code).toMatchSnapshot();
    });

    it('sends a Babel tools object and transformer options to the transformer function', async () => {
      const {code} = await transformFile('transformer-params');
      expect(code).toMatchSnapshot();
    });

    it('sends transformer options and Babel plugin options to a decorator', async () => {
      const {code} = await transformFile('decorator-params');
      expect(code).toMatchSnapshot();
    });

    it('allows to specify options with JSON', async () => {
      const {code} = await transformFile('json', 'options.json');
      expect(code).toMatchSnapshot();
    });

    it('throws an error if filename is not provided', async () => {
      await expect(
        transform('errors', 'options-filename'),
      ).rejects.toThrowError(
        '[AST Decorators]: AST Decorators system requires filename to be set',
      );
    });

    it('throws an error if transformers option is not provided', async () => {
      await expect(
        transform('errors', 'options-transformers'),
      ).rejects.toThrowError('[AST Decorators]: No transformers provided');
    });
  });
});
