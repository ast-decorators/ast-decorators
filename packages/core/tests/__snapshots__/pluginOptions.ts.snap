// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ast-decorators/core plugin-options allows to specify options with JSON 1`] = `
"class Link extends HTMLAnchorElement {
  _bar = 'baz';
}

console.log('bar');"
`;

exports[`@ast-decorators/core plugin-options compiles in another way if option has different state 1`] = `
"class Link extends HTMLAnchorElement {
  _bar = 'baz';
}"
`;

exports[`@ast-decorators/core plugin-options compiles in one way if option has one state 1`] = `
"class Link extends HTMLAnchorElement {
  #_bar = 'baz';
}"
`;

exports[`@ast-decorators/core plugin-options sends a Babel tools object and transformer options to the transformer function 1`] = `
"class Link extends HTMLAnchorElement {
  #_bar = 'baz';
}"
`;

exports[`@ast-decorators/core plugin-options sends transformer options and Babel plugin options to a decorator 1`] = `
"class Link extends HTMLAnchorElement {
  _bar = \\"packages/core/tests/fixtures/plugin-options/decorator-params/input.ts\\";
}"
`;
